# Just for Google Colab

In [ ]:
import os
from getpass import getpass
import urllib

user = input('User name: ')
password = getpass('Password: ')
password = urllib.parse.quote(password) # your password is converted into url format

cmd_string = 'git clone https://{0}:{1}@github.com/GianCarloMilanese/dsim_project.git'.format(user, password)

os.system(cmd_string)
cmd_string, password = "", "" # removing the password from the variable

In [ ]:
!ls -lh dsim_project/

In [ ]:
! git clone https://github.com/Jakobovski/free-spoken-digit-dataset.git && mv free-spoken-digit-dataset/recordings dsim_project/

In [ ]:
!ls -lh dsim_project

In [ ]:
import sys
sys.path.insert(1, "dsim_project/")

# Import libraries

In [1]:
import cnn_models
import data_preparation
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np
from sklearn.svm import SVC
import tensorflow as tf
import data_augmentation
import random

/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/Users/kappa/opt/miniconda3/envs/dsim

# Fix seed

In [2]:
SEED = 10
random.seed(SEED)
tf.random.set_random_seed(SEED)# if working on tf < 2.0
#tf.random.set_seed(SEED)

# Load recordings
## STANDARD RECORDINGS - No spectrogram normalization

In [3]:
fsdd_dir="dsim_project/recordings"
our_recs_dir="dsim_project/preprocessed_recs"

In [4]:
fsdd_dir="./recordings/"
our_recs_dir="./preprocessed_recs/"

In [5]:
recordings = data_preparation.load_recordings(paths=[fsdd_dir, our_recs_dir])

Loading from ./recordings/



Loading from ./preprocessed_recs/


Raw recordings have different lengths? Let's check it out:

In [6]:
min_y = min(map(np.shape, recordings))[0]
max_y = max(map(np.shape, recordings))[0]
print(min_y, max_y)

1010 18262


The difference is quite huge! Let's see which are the longest recordings:

In [7]:
a = [len(x) for x in recordings]
a.sort(reverse=True)
a[0:10]

[18262, 17567, 9015, 8995, 8435, 8281, 8201, 8068, 7755, 7356]

Two recordings have length 18262 and 17567, while the others are around 20K. Let's identify them:

In [8]:
a = [len(x) for x in recordings]
first_length=18262
second_length=17567
index_first = a.index(first_length)
index_second = a.index(second_length)

In [9]:
len(recordings[index_first])

18262

In [10]:
len(recordings[index_second])

17567

I have found them. For knowing to which digit and speaker they are associated I first need to load the labels:

In [11]:
labels_speakers = data_preparation.load_labels(paths=[fsdd_dir, our_recs_dir], label_type="speakers")
labels_digits = data_preparation.load_labels(paths=[fsdd_dir, our_recs_dir])

In [12]:
print("Longest track is associated with speaker {}, digit {}".format(labels_speakers[index_first],labels_digits[index_first]))
print("Second longest track is associated with speaker {}, digit {}".format(labels_speakers[index_second],labels_digits[index_second]))

Longest track is associated with speaker theo, digit 9
Second longest track is associated with speaker theo, digit 7


So the problem is with theo, which has 500 recordings, digit 9 and 7, which respectively have 200 recordings. We can safely delete them and saving to pad many thousands of 0s (there will be (18262 - 9015) less zeros)

In [13]:
max_track_length=17000 # it will be useful later on
print("Before: {}".format(len(recordings)))
recordings=np.delete(recordings,[index_first, index_second])
print("After: {}".format(len(recordings)))

Before: 2400
After: 2398


In [14]:
print("Before: {}".format(len(labels_speakers)))
labels_speakers=np.delete(labels_speakers,[index_first, index_second])
print("After: {}".format(len(labels_speakers)))

Before: 2400
After: 2398


In [15]:
print("Before: {}".format(len(labels_digits)))
labels_digits=np.delete(labels_digits,[index_first, index_second])
print("After: {}".format(len(labels_digits)))

Before: 2400
After: 2398


Let's now double check to see if everything went well. Now the longest recording will be around 9 K

In [16]:
a = [len(x) for x in recordings]
a.sort(reverse=True)
a[0:10]

[9015, 8995, 8435, 8281, 8201, 8068, 7755, 7356, 7147, 7038]

Yes! However the recordings have all different lengths: for this reason we can add 0s at the beginning and at the end in order to uniform them

In [17]:
pad_recordings = data_preparation.pad_zeros(recordings)

pad_zeros >>>
pad_zeros <<<


What is the range now?

In [18]:
min_y = min(map(np.shape, pad_recordings))[0]
max_y = max(map(np.shape, pad_recordings))[0]
print(min_y, max_y)

9015 9015


We can now compute spectograms:

In [19]:
spects = [data_preparation.compute_spectrogram(x) for x in pad_recordings]
spects = np.array(spects)

Let's also compute "normalized spectrograms

In [20]:
norm_spects = [data_preparation.compute_spectrogram(x, normalize=True) for x in pad_recordings]
norm_spects = np.array(norm_spects)

## Augmentation

In [21]:
%%time
X_train_digit, y_train_digit, X_val_digit, y_val_digit, X_test_digit, y_test_digit = data_preparation.prepare_augmented_recordings(audio_dirs= [fsdd_dir, our_recs_dir],
                             y_type= ['digit', 'digit'],
                             n_category_test=15,
                             include_pitch=True,
                             max_length=max_track_length)

split_and_augment_dataset >>>
enrich_dataset>>>
Max length: 17000, shape:(17567,)
Max length: 17000, shape:(18262,)
enrich_dataset <<<
split_and_augment_dataset <<<
split_and_augment_dataset >>>
enrich_dataset>>>
enrich_dataset <<<
split_and_augment_dataset <<<
conversion_done!
compute_spectrograms >>>
9015
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
Padding done
compute_spectrograms <<<
CPU times: user 4min 49s, sys: 10.8 s, total: 5min
Wall time: 3min 55s


In [22]:
print("Lengths : {}, {}, {}, {}, {}, {}".format(len(X_train_digit),
                                                len(y_train_digit),
                                                len(X_val_digit),
                                                len(y_val_digit),
                                                len(X_test_digit),
                                                len(y_test_digit),))

Lengths : 18462, 18462, 4616, 4616, 300, 300


In [23]:
%%time
X_train_speaker, y_train_speaker, X_val_speaker, y_val_speaker, X_test_speaker, y_test_speaker = data_preparation.prepare_augmented_recordings(
    audio_dirs= [our_recs_dir, fsdd_dir],
    y_type= ['speakers_us', 'speakers_default'],
    n_category_test=30,
    include_pitch=True,
    max_length=max_track_length)

split_and_augment_dataset >>>
enrich_dataset>>>
enrich_dataset <<<
split_and_augment_dataset <<<
split_and_augment_dataset >>>
enrich_dataset>>>
Max length: 17000, shape:(17567,)
Max length: 17000, shape:(18262,)
enrich_dataset <<<
split_and_augment_dataset <<<
conversion_done!
compute_spectrograms >>>
9015
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
Padding done
compute_spectrograms <<<
CPU times: user 4min 57s, sys: 13.4 s, total: 5min 10s
Wall time: 4min 5s


In [24]:
print("Lengths : {}, {}, {}, {}, {}, {}".format(len(X_train_speaker),
                                        len(y_train_speaker),
                                        len(X_val_speaker),
                                        len(y_val_speaker),
                                        len(X_test_speaker),
                                        len(y_test_speaker)))

Lengths : 18990, 18990, 4748, 4748, 240, 240


# Standard recordings
## Numbers

Split data in train, val and test

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(spects, labels_digits)

In [ ]:
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")

In [ ]:
%%time
clf1 = clf1.fit(X_train, y_train)

In [ ]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

### Normalize spectrograms

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_digits)

In [ ]:
%%time
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)

In [ ]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

### CNNs

#### Normalized spectrograms

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(norm_spects, labels_digits)

In [ ]:
model = cnn_models.paper_architecture(10, input_shape=input_shape)

In [ ]:
callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss', restore_best_weights=True, patience=3)

In [ ]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

In [ ]:
y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(y_val_nn, y_pred))

#### Standard spectrogram

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(spects, labels_digits)

In [ ]:
model = cnn_models.paper_architecture(10, input_shape)

In [ ]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

In [ ]:
y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(y_val_nn, y_pred))

From what we can see normalising spectrograms is the way to go. Let's use it by default

### Best model
The best model for this task is the paper CNN, using the "normalized" spectrograms. Let's train it on X_train + X_val for 10 epochs and evaluate it on the not-yet-seen test set

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(norm_spects, labels_digits)
model = cnn_models.paper_architecture(10, input_shape=input_shape)

In [ ]:
X_train = np.concatenate([X_train, X_val], axis=0)
y_train = np.concatenate([y_train, y_val], axis=0)

In [ ]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1)

In [ ]:
y_test = np.argmax(y_test, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_test)
print(classification_report(y_test, y_pred))

## Speakers
### SVD

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_speakers)

In [ ]:
%%time
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)

In [ ]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

### CNN

For neural networks it is not possible to pass the labels as-is: we need to transform them in numbers. The safest way is through one-hot encoding

In [ ]:
y, target_names = data_preparation.transform_categorical_y(labels_speakers)

In [ ]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(norm_spects, y, number_mode=False)

In [ ]:
model = cnn_models.paper_architecture(8, input_shape=input_shape)

In [ ]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

In [ ]:
Y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(Y_val_nn, y_pred))

#### Paper - batch_normalisation=True

In [ ]:
model = cnn_models.paper_architecture(8, input_shape, batch_normalisation=True)

In [ ]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

In [ ]:
y_pred = model.predict_classes(X_val)
print(classification_report(Y_val_nn, y_pred))

### Best model

In [ ]:
%%time
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_speakers)
X_train = np.concatenate([X_train, X_val], axis=0)
y_train = np.concatenate([y_train, y_val], axis=0)
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)
y_pred = clf1.predict(X_test)
print(classification_report(y_test, y_pred))

# Data augmentation
## Speaker

In [ ]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder()
y_train_speaker = enc.fit_transform(np.array(y_train_speaker).reshape(-1, 1)).toarray()
y_val_speaker = enc.transform(np.array(y_val_speaker).reshape(-1, 1)).toarray()
y_test_speaker = enc.transform(np.array(y_test_speaker).reshape(-1, 1)).toarray()
label_0 = enc.inverse_transform(np.array([1, 0, 0, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_1 = enc.inverse_transform(np.array([0, 1, 0, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_2 = enc.inverse_transform(np.array([0, 0, 1, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_3 = enc.inverse_transform(np.array([0, 0, 0, 1, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_4 = enc.inverse_transform(np.array([0, 0, 0, 0, 1, 0, 0, 0]).reshape(1, -1))[0][0]
label_5 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 1, 0, 0]).reshape(1, -1))[0][0]
label_6 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 0, 1, 0]).reshape(1, -1))[0][0]
label_7 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 0, 0, 1]).reshape(1, -1))[0][0]
target_names = [label_0, label_1, label_2, label_3, label_4, label_5, label_6, label_7]

In [ ]:
X_train_speaker = np.array(X_train_speaker)
X_val_speaker = np.array(X_val_speaker)
X_test_speaker = np.array(X_test_speaker)

In [ ]:
X_train_speaker = X_train_speaker.reshape(X_train_speaker.shape[0],
                                          X_train_speaker.shape[1],
                                          X_train_speaker.shape[2],
                                          1)
X_val_speaker = X_val_speaker.reshape(X_val_speaker.shape[0],
                                      X_val_speaker.shape[1],
                                      X_val_speaker.shape[2],
                                      1)
X_test_speaker = X_test_speaker.reshape(X_test_speaker.shape[0],
                                        X_test_speaker.shape[1],
                                        X_test_speaker.shape[2],
                                        1)

In [ ]:
input_shape = (X_train_speaker.shape[1], X_train_speaker.shape[2], 1)

In [ ]:
model = cnn_models.paper_architecture(8, input_shape=input_shape)

In [ ]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

In [ ]:
Y_val_nn = np.argmax(y_val_speaker, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

### Batch_normalization = True

In [ ]:
model = cnn_models.paper_architecture(8, input_shape=input_shape, batch_normalisation=True)

In [ ]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

In [ ]:
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

### Different architecture
Let's change a bit the architecture and see if we can improve scores:

In [ ]:
model = cnn_models.custom_cnn(8, input_shape=input_shape)

In [ ]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

In [ ]:
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

### Best model
Based on the f1-score, the best model is the "custom cnn" one. Let's see its result on the test set:

In [ ]:
%%time
X_train = np.concatenate([X_train_speaker, X_val_speaker], axis=0)
y_train = np.concatenate([y_train_speaker, y_val_speaker], axis=0)
model = cnn_models.custom_cnn(8, input_shape=input_shape)
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1)

In [ ]:
y_test_nn = np.argmax(y_test_speaker, axis=1)
y_pred = model.predict_classes(X_test_speaker)
print(classification_report(y_test_nn, y_pred, target_names=target_names))

### Digits

In [ ]:
X_train_digit_nn = np.array(X_train_digit)
X_val_digit_nn = np.array(X_val_digit)
X_test_digit_nn = np.array(X_test_digit)

In [ ]:
X_train_digit_nn = X_train_digit_nn.reshape(X_train_digit_nn.shape[0], X_train_digit_nn.shape[1], X_train_digit_nn.shape[2], 1)
X_val_digit_nn = X_val_digit_nn.reshape(X_val_digit_nn.shape[0], X_val_digit_nn.shape[1], X_val_digit_nn.shape[2], 1)
X_test_digit_nn = X_test_digit_nn.reshape(X_test_digit_nn.shape[0], X_test_digit_nn.shape[1], X_test_digit_nn.shape[2], 1)
y_train_digit_nn = tf.keras.utils.to_categorical(y_train_digit, 10)
y_test_digit_nn = tf.keras.utils.to_categorical(y_test_digit, 10)

In [ ]:
y_val_digit_nn = tf.keras.utils.to_categorical(y_val_digit, 10)

In [ ]:
input_shape = (X_train_digit_nn.shape[1], X_train_digit_nn.shape[2], 1)

#### Paper

In [ ]:
model = cnn_models.paper_architecture(10, input_shape=input_shape, batch_normalisation=True)

In [ ]:
%%time
model.fit(X_train_digit_nn, y_train_digit_nn,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_digit_nn, y_val_digit_nn))

In [ ]:
Y_val_nn = np.argmax(y_val_digit_nn, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val_digit_nn)
print(classification_report(Y_val_nn, y_pred))

#### Custom

In [ ]:
model = cnn_models.custom_cnn(10, input_shape=input_shape)

In [ ]:
%%time
model.fit(X_train_digit_nn, y_train_digit_nn,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_digit_nn, y_val_digit_nn))

In [ ]:
y_pred = model.predict_classes(X_val_digit_nn)
print(classification_report(Y_val_nn, y_pred))

### Best model
Based on F1-Score the best model is once again the custom paper architecture:

In [ ]:
%%time
X_train = np.concatenate([X_train_digit_nn, X_val_digit_nn], axis=0)
y_train = np.concatenate([y_train_digit_nn, y_val_digit_nn], axis=0)
model = cnn_models.custom_cnn(10, input_shape=input_shape)
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1)

In [ ]:
y_test_nn = np.argmax(y_test_digit_nn, axis=1)
y_pred = model.predict_classes(X_test_digit_nn)
print(classification_report(y_test_nn, y_pred))

# Test model 

In [ ]:
import sounddevice as sd
import subprocess

import time
import librosa

import IPython.display as ipd

import os
from scipy.io import wavfile as wav

In [ ]:
def pad_zeros_single_rec(rec, max_y):
    rec = np.array(rec)
    diff_in_y = max_y - rec.shape[0]
    if diff_in_y > 0:
        half_diff = int(diff_in_y/2)
        remaining_diff = diff_in_y-half_diff
        v = np.pad(rec, (half_diff, remaining_diff), 'constant', constant_values=0)
        return v
    else:
        return rec

In [ ]:
def create_recording(duration, rec_rate, name = "test.wav", output_dir = "test/"):
    print("Ready in 3...", end = "")
    time.sleep(1)
    print("2...", end = "")
    time.sleep(1)
    print("1...")
    time.sleep(1)
    print("Go.")
    rec = sd.rec(int(duration * rec_rate), samplerate=rec_rate, channels=1, blocking=True)
    print("Playing the recording.")
    sd.play(rec, rec_rate)

    # after hearing the recording, decide whether to record it again or continue to next number
    # if you type anything, record again
    # if you press enter, save current recording & go to next number
    ok = input("OK?")
    if ok == "":
        librosa.output.write_wav(output_dir+name, rec, rec_rate)
        return rec
    ipd.clear_output(wait=True)
    create_recording(duration, rec_rate)

In [ ]:
def trim_audio(file, input_dir="test/", output_dir="test/", db=-48):

    if not os.path.isdir(input_dir):
        print(f"There should be an input \"{input_dir}\" directory.")
        sys.exit(0)
    
    # create output directory if not there yet
    if not os.path.isdir(output_dir):
        os.makedirs(output_dir)
        
    temp1 = output_dir+"temp1.wav"
    temp2 = output_dir+"temp2.wav"
    temp3 = output_dir+"temp3.wav"
 
    subprocess.run(["ffmpeg", "-y", "-i", input_dir+file, "-af", f"silenceremove=1:0:{db}dB", temp1])
    subprocess.run(["ffmpeg", "-y", "-i", temp1, "-af", "areverse", temp2])
    subprocess.run(["ffmpeg", "-y", "-i", temp2, "-af", f"silenceremove=1:0.1:{db}dB", temp3])
    subprocess.run(["ffmpeg", "-y", "-i", temp3, "-af", "areverse", output_dir+file])
    
    os.remove(temp1)
    os.remove(temp2)
    os.remove(temp3)

In [ ]:
def test_NN(nn, max_y, target_names, answer = None, duration=2, rec_rate=8000, directory = "test/", filename = "test.wav"):
    create_recording(duration, rec_rate, filename, directory)   
    ipd.clear_output()
    trim_audio(filename, directory, directory)
    # _, rec = wav.read(directory + "/" + filename)
    rec, _ = librosa.core.load(directory + "/" + filename, sr = rec_rate)
    rec = pad_zeros_single_rec(rec, max_y)
    # sd.play(rec, rec_rate)
    rec = data_preparation.compute_spectrogram(rec, normalize=True)
    rec = rec[np.newaxis,:,:,np.newaxis]
    preds = nn.predict_classes(rec)
    print("Model prediction: {}".format(target_names[preds[0]]))
    if answer is not None:
        print(f"Correct answer {answer}")
    return preds

In [ ]:
max_y = len(data_augm_pad_recordings[1])

In [ ]:
pred = test_NN(model, max_y, target_names, answer = "gian")

# TO DO:
- [x] Set random seed
- [x] Use only original recordings in test set of augmented scenario
- [x] Use proper validation set for picking best models and params
- [x] Data augmentation also for digit recognition
- [ ] Evaluate each best model on test set, after training it on x_train + x_test